Cookbook/Proposals/ComposableReducers.md

```diff
@@ -0,0 +1,250 @@
+# Composable Reducers using Lenses and Prisms
+
+* Author: Yasuhiro Inami
+* Review Manager: David Rodrigues
+
+## Introduction
+
+This proposal adds functional [`Lens` and `Prism`](#functional-lens--prism) to break down `reducer` / `state` / `event` (`action`) in [ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) so that we can create a large application from small component compositions.
+
+## Motivation
+
+In Babylon.app, we are using [ReactiveFeedback](https://github.com/Babylonpartners/ReactiveFeedback) to control states, side-effects, and feedback loops to define a particular screen (view controller) behavior, owned by `ViewModel`.
+
+However, this `ViewModel` can easily become too complex as `reducer: (State, Event) -> State` grows large with tons of pattern-matching.
+
+Unfortunately, splitting into multiple `ViewModel`s is not a clever solution, as managing multiple `ReactiveFeedback`s that interact with each other tends to be hard to control.
+
+To curcumvent this problem, we instead **split `reducer`s and also `state`s and `event`s**, and then combine them using `Lens` and `Prism`.
+
+The basic idea can be found in @mbrandonw ’s talk:
+
+[Brandon Williams \- Composable Reducers & Effects Systems \- YouTube](https://www.youtube.com/watch?v=QOIigosUNGU)
+
+## Proposed solution
+
+### Functional `Lens` & `Prism`
+
+```swift
+/// For accessing struct members.
+/// e.g. Whole = whole struct (members), Part = partial member
+struct Lens<Whole, Part> {
+    let get: (Whole) -> Part
+    let set: (Whole, Part) -> Whole
+}
+
+/// For accessing enum cases.
+/// e.g. Whole = all possible enum cases, B = partial case
+struct Prism<Whole, Part> {
+    let preview: (Whole) -> Part?
+    let review: (Part) -> Whole
+}
+```
```

Comment:
Please add an example of its usage.

Reply:
cf. https://speakerdeck.com/inamiy/making-your-own-code-formatter-in-swift?slide=41

`Prism` is useful when traversing some tree structure (that is `struct`) but its descendant is `Optional` (which is `enum` ). 
Just using plain `Lens` composition can't achieve `Optional`'s getter and setter (because it may fail), and we want `Prism` layer to reinforce the traversal.

Do you think this will work for the proposal documentation?

Reply:
9122ca4 Added "Why Prism?" section as example.